# Bioinformatic Analysis in the SDG lab!

Hello and welcome to your crash course in bioinformatic analysis in the lab! Regardless of your coding experience if you're interested in learning how to analyse your own data here are some handy guides. 

I recommend following the guides in order - any questions feel free to send me a message. 

## Imperial Courses
If you have no previous coding experience, this can feel a bit daunting, but think about it like learning a new language. We will mainly be using three programming langauges, Command Line, R and Python. I would highly recommend making use of [Imperial's data science courses](https://www.imperial.ac.uk/students/academic-support/graduate-school/professional-development/doctoral-students/research-computing-data-science/courses/) - these are a great resource and will give you a kickstart in understanding basic coding language and help you write your first scripts so you can understand more complicated scripts in the future. I'd highly recommed completing at least:

- [Basic statistics](https://www.imperial.ac.uk/students/academic-support/graduate-school/professional-development/doctoral-students/research-computing-data-science/courses/basic-statistics/)
- [Introduction to R](https://www.imperial.ac.uk/students/academic-support/graduate-school/professional-development/doctoral-students/research-computing-data-science/courses/r-programming/)
- [Data processing with R](https://www.imperial.ac.uk/students/academic-support/graduate-school/professional-development/doctoral-students/research-computing-data-science/courses/data-processing-with-r/)
- [Introduction to Python](https://www.imperial.ac.uk/students/academic-support/graduate-school/professional-development/doctoral-students/research-computing-data-science/courses/intro-to-python/)
- [The Linux command line for scientific computing](https://www.imperial.ac.uk/students/academic-support/graduate-school/professional-development/doctoral-students/research-computing-data-science/courses/linux-command-line-for-scientific-computing/)
- [Introduction to HPC](https://www.imperial.ac.uk/students/academic-support/graduate-school/professional-development/doctoral-students/research-computing-data-science/courses/introduction-to-hpc/)
- [Data exploration and visualisation](https://www.imperial.ac.uk/students/academic-support/graduate-school/professional-development/doctoral-students/research-computing-data-science/courses/data-exploration-visualisation/)
- [Introduction to Git and Github for software development](https://www.imperial.ac.uk/students/academic-support/graduate-school/professional-development/doctoral-students/research-computing-data-science/courses/introduction-to-using-git-and-github-for-software-development/)


These will give you the fundamentals from which you can build on, if you are interested in learning more. Make sure to book these as soon as possible because the spaces get filled very quickly. 

## Other resources
- [Code Academy](https://www.codecademy.com) They have many courses, but they also have docs, cheat sheets, and much more. 
- [Biostars](https://www.biostars.org) is great for any questions you have or errors you come across that you're struggling to solve. Or if you want to write a code in a particular way, usually you'll find that people have already written somehting similar, which makes it easier for you. 
- [Bioconductor](https://bioconductor.org) for R specific coding. The majority of the packages that we use are from Bioconductor, as this is the main R-specific site for data analysis. 
- [SEQanswers](https://www.seqanswers.com/forum/bioinformatics/bioinformatics-aa/70310-how-to-add-gfp-sequence-to-mouse-genome-for-mapping) for sequencing specific questions.  
- [Coursera](https://www.coursera.org) or [Edx](https://www.edx.org) for free courses. 
- [ChatGPT](https://openai.com/blog/chatgpt) obviously - can help you find errors in your code, write code snippets for you etc. You'll need to make an account to use it. 








